File: webui/src/main/java/com/taskadapter/webui/ConfigureSystemPage.java
package com.taskadapter.webui;

import com.taskadapter.auth.AuthorizedOperations;
import com.taskadapter.auth.CredentialsManager;
import com.taskadapter.license.License;
import com.taskadapter.web.SettingsManager;
import com.taskadapter.webui.user.UsersPanel;
import com.vaadin.data.Property;
import com.vaadin.ui.Alignment;
import com.vaadin.ui.CheckBox;
import com.vaadin.ui.Component;
import com.vaadin.ui.Panel;
import com.vaadin.ui.VerticalLayout;

import static com.vaadin.server.Sizeable.Unit.PIXELS;

public final class ConfigureSystemPage {

    private static Component createAdminPermissionsSection(
            final SettingsManager settingsManager, boolean modifiable) {
        final Panel panel = new Panel();
        final VerticalLayout view = new VerticalLayout();
        view.setMargin(true);
        panel.setContent(view);

        final CheckBox checkbox = new CheckBox(
                "Admin can view and manage all users' configs");
        checkbox.setValue(settingsManager.adminCanManageAllConfigs());
        checkbox.setImmediate(true);
        checkbox.addValueChangeListener(new CheckBox.ValueChangeListener() {
            @Override
            public void valueChange(Property.ValueChangeEvent valueChangeEvent) {
                settingsManager.setAdminCanManageAllConfigs(checkbox.getValue());
            }
        });
        view.addComponent(checkbox);
        view.setComponentAlignment(checkbox, Alignment.MIDDLE_LEFT);
        checkbox.setEnabled(modifiable);

        return panel;
    }

    public static Component render(CredentialsManager credentialsManager,
            SettingsManager settings, License license,
            AuthorizedOperations authorizedOps) {
        final VerticalLayout layout = new VerticalLayout();

        layout.setSpacing(true);
        final Component cmt = LocalRemoteOptionsPanel.createLocalRemoteOptions(
                settings, authorizedOps.canConfigureServer());
        cmt.setWidth(500, PIXELS);
        layout.addComponent(cmt);

        layout.addComponent(createAdminPermissionsSection(settings,
                authorizedOps.canConfigureServer()));

        layout.addComponent(UsersPanel.render(credentialsManager,
                authorizedOps, license));

        return layout;

    }
}

File: webui/src/main/scala/com/taskadapter/webui/ConfigureSystemPage.scala
package com.taskadapter.webui

import com.taskadapter.auth.{AuthorizedOperations, CredentialsManager}
import com.taskadapter.license.License
import com.taskadapter.web.SettingsManager
import com.taskadapter.web.service.Sandbox
import com.taskadapter.webui.pages.Navigator
import com.taskadapter.webui.service.Preservices
import com.taskadapter.webui.user.UsersPanel
import com.vaadin.flow.component.{Component, HasComponents}
import com.vaadin.flow.component.checkbox.Checkbox
import com.vaadin.flow.component.dependency.CssImport
import com.vaadin.flow.component.formlayout.FormLayout
import com.vaadin.flow.component.orderedlayout.VerticalLayout
import com.vaadin.flow.component.textfield.TextField
import com.vaadin.flow.router.Route

@Route(value = Navigator.CONFIGURE_SYSTEM, layout = classOf[Layout])
@CssImport(value = "./styles/views/mytheme.css")
class ConfigureSystemPage() extends BasePage {

  private val configOps: ConfigOperations = SessionController.buildConfigOperations()
  private val services: Preservices = SessionController.getServices

  render(services.credentialsManager,
    services.settingsManager,
    services.licenseManager.getLicense,
    configOps.authorizedOps)

  private def createAdminPermissionsSection(settingsManager: SettingsManager, modifiable: Boolean) = {
    val view = new VerticalLayout
    addCheckbox(view, Page.message("configurePage.showAllUsersConfigs"), settingsManager.schedulerEnabled, modifiable,
      newValue => settingsManager.setSchedulerEnabled(newValue))

    addCheckbox(view, Page.message("configurePage.anonymousErrorReporting"), settingsManager.isErrorReportingEnabled, modifiable,
      newValue => settingsManager.setErrorReporting(newValue))

    view
  }

  private def addCheckbox(view: VerticalLayout, label: String, value: Boolean, modifiable: Boolean, listener: Boolean => Unit) = {
    val checkbox = new Checkbox(label)
    checkbox.setValue(value)
//    checkbox.setImmediate(true)
    checkbox.addValueChangeListener(_ => listener.apply(checkbox.getValue))
    view.add(checkbox)
//    view.setComponentAlignment(checkbox, Alignment.MIDDLE_LEFT)
    checkbox.setEnabled(modifiable)
  }

  def createResultsNumberSection(settingsManager: SettingsManager): Component = {
    val view = new VerticalLayout
    val field = new TextField(Page.message("configurePage.maxNumberOfResultsToSave"))
//    field.setDescription(Page.message("configurePage.maxNumberExplanation"))
    field.setValue(settingsManager.getMaxNumberOfResultsToKeep + "")
    field.addValueChangeListener(_ => settingsManager.setMaxNumberOfResultsToKeep(field.getValue.toInt))
    view.add(field)
    view
  }

  def render(credentialsManager: CredentialsManager, settingsManager: SettingsManager, license: License,
             authorizedOps: AuthorizedOperations): Unit = {

    setSpacing(true)
    val cmt = LocalRemoteOptionsPanel.createLocalRemoteOptions(settingsManager, authorizedOps.canConfigureServer)
    add(cmt)
    val allowedToEdit = authorizedOps.canConfigureServer && license != null
    add(createAdminPermissionsSection(settingsManager, allowedToEdit))
    add(createResultsNumberSection(settingsManager))
    add(new UsersPanel(credentialsManager, authorizedOps, license))

  }
}
